README.md
#                Programmer's Calculator  (Ver 1.5)

##  Arithmetic in base 16, base 10, base 2 (hex, dec, bin)

 Resurrected this because it may be useful. For instance, in embedded programming it
is nice to see results from mixed operations, - for example - like '0xA0000+27*2'.
(The address of 27th char in the framebuffer)

 For some, it may  serve as a good introduction to LEX/YACC.

 Came to learn, that this code was used in several commercial products, which
I welcome with open arms, even if no credit was given. Glad I could make an impact.

  I got to find out that one of the products - based upon this code - is successful
enough to maintain a nice farm. Would be ice to have a side/guest house there to visit.

## Tech:

      Operators line in most languages. Precedence like in 'C'.

 Operator summary:

        right  assotiation:     '='         ASSIGNMENT\n\
        left   assotiation:     or  and     BINARY OR/BINARY AND\n\
        left   assotiation:     '+' '-'     ADDITION/SUBTRUCTION\n\
        left   assotiation:     '<<' '>>'   SHIFT LEFT / RIGHT (Integer)\n\
        left   assotiation:     '*' '/' '%' MULTIPLY/DIVIDE/MODULUS\n\
        left   assotiation:     '-'         UNARY MINUS\n\
        right  assotiation:     '^'         EXPONENTATION\n\

 Math functions summary:

        SIN         math sinus              COS        math cos
        ATAN        math atan               LOG        math log
        LOG10       math log10              EXP        math exp
        SQRT        math sqrt               INT        math int
        ABS         math abs

 Function summary:

        F2C         converting temperature -> fahrenheit to celsius
        C2F         converting temperature -> celsius to fahrenheit
        IN2MM       converting length      -> inches to millimeter
        MM2IN       converting length      -> millimeter to inches
        PO2KG       converting mass        -> pounds to kilogram
        KG2PO       converting mass        -> kilogram to pounds

 Constans summary:

        PI      3.14159265358979323846
        E       2.71828182845904523536
        GAMMA   0.57721566490153286060
        DEG     57.29577951308322087680
        PHI     1.61803389974989484820

Type pcalc with no arguments for help summary.

See the file INSTALL for building and installation instructions.
(the usual configure / make )

## Copyright

   pcalc is free software.

## Command line usage summary

 The command line parser pre-parses the strings you pass on. If the
command line is not delivering to pcalc what you expect, you may
enclose the command in quotes; For instance the string 2 / 3 is
parsed into three arguments, so using "2 / 3" delivers what you
intended. As a side note, the string 2/3 is interpreted as
you intended.

Peter Glen







NEWS
             			    PCALC

pcalc is decades old, I resurrected it as it is a useful tool even in modern computing.

xx.xx.94            -- first code
nov 26 2019	        -- put it on github
Mon 25.Jul.2022     -- added shift operator
Thu 22.Dec.2022     -- moved to src dir

Free to use;


symbol.c
/*
 *  pcalc
 */

#include <stdio.h>
#include <malloc.h>
#include <string.h>
#include <math.h>

#include "hocdecl.h"
#include "symbol.h"
#include "pcalc.h"

static  Symbol *symlist = (Symbol *) 0 ;

typedef struct
        {
        char *name ;
        double cval;
        }
        CONSTS;

CONSTS consts[] =

{
        "PI",    3.14159265358979323846,
        "E",     2.71828182845904523536,
        "GAMMA", 0.57721566490153286060,
        "DEG",   57.29577951308322087680,
        "PHI",   1.61803389974989484820,
        NULL,    0.0
} ;

typedef struct
        {
        char     *name ;
        double   (*func)() ;
        }
        BUILTINS;

BUILTINS    builtins[] =

{
        "sin",      sin,
        "cos",      cos,
        "atan",     atan,
        "log",      Log,
        "log10",    Log10,
        "exp",      Exp,
        "sqrt",     Sqrt,
        "int",      integer,
        "abs",      fabs,
        "f2c",      f2c,
        "c2f",      c2f,
        "in2mm",    in2mm,
        "mm2in",    mm2in,
        "po2kg",    po2kg,
        "kg2po",    kg2po,

        "SIN",      sin,
        "COS",      cos,
        "ATAN",     atan,
        "LOG",      Log,
        "LOG10",    Log10,
        "EXP",      Exp,
        "SQRT",     Sqrt,
        "INT",      integer,
        "ABS",      fabs,
        "F2C",      f2c,
        "C2F",      c2f,
        "IN2MM",    in2mm,
        "MM2IN",    mm2in,
        "PO2KG",    po2kg,
        "KG2PO",    kg2po,

        NULL,       (void *) 0
} ;

typedef struct
        {
        char     *name ;
        int      (*ifunc)() ;
        }
        IBUILTINS;

IBUILTINS    ibuiltins[] =

{

        "date",     ddate,
        "print",    print,
        "echo",     echo,
        "_echo",    echo_nl,

        "DATE",     ddate,
        "PRINT",    print,
        "ECHO",     echo,
        "_ECHO",    echo_nl,

        NULL,       (void *) 0
} ;



void    init_sym(void)

{
    int i ;
    Symbol *s ;

    for (i=0 ; consts[i].name ; i++)
      (void) install_sym( consts[i].name, VAR, consts[i].cval) ;

    for (i=0 ; builtins[i].name ; i++)
        {
        s = install_sym( builtins[i].name, BUILTIN, 0.0) ;
        s->u.ptr = builtins[i].func ;
        }

    for (i=0 ; ibuiltins[i].name ; i++)
        {
        s = install_sym( ibuiltins[i].name, IBUILTIN, 0) ;
        s->u.iptr = ibuiltins[i].ifunc ;
        }
}

void    dump_sym(void)

{
    Symbol *sp ;

    for (sp = symlist ; sp != (Symbol *) 0 ; sp = sp->next)
        {
        printf("Symbol: %s\n", sp->name);
        }
}

/*--------------------------------------------------------------------------
**  LOOKUP  -   find symbol in table
*/

/*
 * find s in symbol table
 */

Symbol  *lookup_sym(char *s)

{
    Symbol *sp ;

    for (sp = symlist ; sp != (Symbol *) 0 ; sp = sp->next)
      if (strcmp( sp->name, s) == 0)
         return sp ;                                             /* found  */

   return (Symbol *) 0 ;                               /* symbol not found */
}

/*--------------------------------------------------------------------------
**  INSTALL -   install symbol in table
*/

Symbol  *install_sym(char *s, int t, double d)

{
    Symbol *sp ;

    sp = (Symbol *) emalloc( sizeof( Symbol)) ;

    sp->name = emalloc( strlen(s) + 1) ;
    strcpy( sp->name, s) ;

    sp->type = t ;
    sp->u.val = d ;
    sp->next = symlist ;                    /* add symbol to front of list */
    symlist = sp ;

    return sp ;
}

/*--------------------------------------------------------------------------
**  EMALLOC -   allocate memory for string, complain on error.
*/

char    *emalloc(unsigned n)

{
 char *p ;

   p = (char *) malloc( n) ;
   if (p == (char *)0)
      execerror("out of memory", (char *) 0) ;
   return p ;
}


/* end of symbol.c */
